ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -126,75 +135,54 @@ public PersistentIndex(String datadir, ScheduledExecutorService scheduler, Log l
-      StoreMetrics metrics, Journal journal, Time time)
+  PersistentIndex(String datadir, ScheduledExecutorService scheduler, Log log, StoreConfig config,
+      StoreKeyFactory factory, MessageStoreRecovery recovery, MessageStoreHardDelete hardDelete, StoreMetrics metrics,
+      Journal journal, Time time)
       throws StoreException {
+    File[] indexFiles = new File(datadir).listFiles(INDEX_FILE_FILTER);
+    if (indexFiles == null) {
+      throw new StoreException("Could not read index files from directory [" + datadir + "]",
+          StoreErrorCodes.Index_Creation_Failure);
+    }
     try {
+      this.dataDir = datadir;
+      this.log = log;
       this.time = time;
       this.scheduler = scheduler;
       this.metrics = metrics;
-      logSegment = log.getFirstSegment();
-      File indexDir = new File(datadir);
-      File[] indexFiles = indexDir.listFiles(new IndexFilter());
       this.factory = factory;
       this.config = config;
-      persistor = new IndexPersistor();
-      hardDeleter = new HardDeleter(config, metrics, datadir, log, this, hardDelete, factory, time);
       this.hardDelete = hardDelete;
       this.journal = journal;
-      Arrays.sort(indexFiles, new Comparator<File>() {
-        @Override
-        public int compare(File o1, File o2) {
-          if (o1 == null || o2 == null) {
-            throw new NullPointerException("arguments to compare two files is null");
-          }
-          // File name pattern for index is offset_name. We extract the offset from
-          // name to compare
-          int o1Index = o1.getName().indexOf("_", 0);
-          long o1Offset = Long.parseLong(o1.getName().substring(0, o1Index));
-          int o2Index = o2.getName().indexOf("_", 0);
-          long o2Offset = Long.parseLong(o2.getName().substring(0, o2Index));
-          if (o1Offset == o2Offset) {
-            return 0;
-          } else if (o1Offset < o2Offset) {
-            return -1;
-          } else {
-            return 1;
-          }
-        }
-      });
+      this.maxInMemoryIndexSizeInBytes = config.storeIndexMaxMemorySizeBytes;
+      this.maxInMemoryNumElements = config.storeIndexMaxNumberOfInmemElements;
+
+      logAbsoluteZeroOffset = new Offset(log.getFirstSegment().getName(), 0);
+      hardDeleter = new HardDeleter(config, metrics, datadir, log, this, hardDelete, factory, time);
 
+      Arrays.sort(indexFiles, INDEX_FILE_COMPARATOR);
       for (int i = 0; i < indexFiles.length; i++) {
-        boolean map = false;
         // We map all the indexes except the most recent index segment.
         // The recent index segment would go through recovery after they have been
         // read into memory
-        if (i < indexFiles.length - 1) {
-          map = true;
-        }
+        boolean map = i < indexFiles.length - 1;
         IndexSegment info = new IndexSegment(indexFiles[i], map, factory, config, metrics, journal);
```

Comment:
Trying to keep unrelated changes out of this PR - already quite large. The index segment is inconsistently named across the `PersistentIndex`. If we are concerned about consistency, we should make those changes as a separate commit.